README.rst
===========
Project status
===========

This is a fork of the unmaintaned upstream `django-suit <https://github.com/darklow/django-suit>`_ project. The `develop` branch is compatible with the latest Django versions, other branches were left as-is on the moment of forking.

Install this using pip and direct github, it is not published on pypi as separate package currently:

```
pip install https://github.com/mhindery/django-suit/archive/develop.zip
```

===========
Django Suit
===========

**Modern theme for Django admin interface**.

Django Suit is alternative theme/skin/extension for `Django <http://www.djangoproject.com>`_ administration interface.

* Project home: http://djangosuit.com/
* Live demo: http://djangosuit.com/admin/


License
=======

* Django Suit is licensed under `Creative Commons Attribution-NonCommercial 3.0 <http://creativecommons.org/licenses/by-nc/3.0/>`_ license.
* Licence and pricing: http://djangosuit.com/pricing/


Docs & Support
==============

* Documentation: http://django-suit.readthedocs.org/
* Support: http://djangosuit.com/support/
* `Demo app <http://djangosuit.com/admin/>`_ repository is here: `django-suit-examples <https://github.com/darklow/django-suit-examples>`_
* IRC Channel: Feel free to join us at ``#django-suit`` on ``irc.freenode.net`` for a chat
* Follow `on Twitter <http://twitter.com/DjangoSuit>`_ to get latest news

Changelog
=========

See `CHANGELOG.rst <https://github.com/darklow/django-suit/blob/develop/CHANGELOG.rst>`_.


Contributing
============

See `Contributing documentation <http://django-suit.readthedocs.org/en/develop/contributing.html>`_


Build Status
============

Django Suit uses Travis CI to perform tests on different Django and Python versions.

Tested using Python: 2.5-3.4. Django: 1.4-1.10. Develop / Master:

.. |develop| image:: https://travis-ci.org/darklow/django-suit.png?branch=develop
   :alt: Build Status - develop branch
   :target: http://travis-ci.org/darklow/django-suit

.. |master| image:: https://travis-ci.org/darklow/django-suit.png?branch=master
   :alt: Build Status - master branch
   :target: http://travis-ci.org/darklow/django-suit

|develop| |master|


Preview
=======

.. image:: https://raw.github.com/darklow/django-suit/develop/docs/_static/img/django-suit.png
   :alt: Django Suit Preview
   :target: http://djangosuit.com/admin/


v2.0-dev
========

Check out new v2.0-dev version notes: https://github.com/darklow/django-suit/issues/475


CHANGELOG.rst
Changelog
=========

Only important changes are mentioned below. See `commit log <https://github.com/darklow/django-suit/commits/develop>`_, `closed issues <https://github.com/darklow/django-suit/issues?direction=desc&sort=updated&state=closed>`_ and `closed pull
requests <https://github.com/darklow/django-suit/pulls?q=sort%3Aupdated-desc+is%3Apr+is%3Aclosed>`_ for full changes.

v.0.3.1 (2022-09-01)

* [Fix] Django 4 compatibility fixes. Thanks to @jorenham for `#1 <https://github.com/mhindery/django-suit/pull/1>`_.

v.0.2.29 (2021-04-28)
--------------------
* This is a 'maintenance fork' which contains some fixes for compatibility with the latest Django releases.


v0.2.28 (2019-10-02)
--------------------

* [Fix] Django 2+ compatibility fixes.
* [Fix] Fix menu not being seen for newest Chrome.
* [Fix] Fixes by PR.


v0.2.26 (2018-02-24)
--------------------

* [Fix] Django 2.0 support.
* [Fix] Minor fixes by PR.


v0.2.25 (2017-03-30)
--------------------

* [Fix] Fixes missing original time/date widget icons.
* [Fix] Fixes missing menu when using with Django Channels.


v0.2.24 (2017-01-27)
--------------------

* [Fix] Sortable inlines required fields detection fix (related to jQuery update).
* [Fix] More precise SPAN selector not to mess up editors and other content in stacked inlines.


v0.2.23 (2016-12-06)
--------------------

* [Fix] Fieldset newlines fix, jQuery tab fix and few other minor bugs by PR.


v0.2.22 (2016-11-11)
--------------------

* [Fix] Upgrade to jQuery 2.2.4 when using Django >= 1.9


v0.2.21 (2016-08-18)
--------------------

* [Fix] Django 1.10 compatibility and Travis tests fixes


v0.2.20 (2016-08-09)
--------------------

* [Fix] Django 1.10 compatibility fixes `#526 <https://github.com/darklow/django-suit/pull/526>`_


v0.2.19 (2016-06-23)
--------------------

* [Fix] Multiple fixes by Pull Requests


v0.2.18 (2016-03-03)
--------------------

* [Fix] Django 1.9 time and date widgets fixes `#481 <https://github.com/darklow/django-suit/issues/481>`_


v0.2.18 (2016-03-03)
--------------------

* [Fix] Django 1.9 time and date widgets fixes `#481 <https://github.com/darklow/django-suit/issues/481>`_


v0.2.17 (2016-02-23)
--------------------

* [Fix] Django 1.9 and Python 3 Compatibility fixes.


v0.2.16 (2016-01-16)
--------------------

* [Fix] Django 1.9 Compatibility fixes.


v0.2.15 (2015-09-28)
--------------------

* [Fix] Multiple Django 1.8/1.9/1.10 DeprecationWarning and compatibility fixes.
* [Fix] Fixed bug with not showing login errors in Django>1.8 `#410 <https://github.com/darklow/django-suit/pull/410>`_
* [Fix] Menu fixes and improvements `#430 <https://github.com/darklow/django-suit/pull/430>`_
* [Fix] Reversion app support fixes `#424 <https://github.com/darklow/django-suit/pull/424>`_
* [Feature] New `before_suit_styles` block in base.html `#408 <https://github.com/darklow/django-suit/pull/408>`_


v0.2.14 (2015-07-23)
--------------------

* [Fix] Django 1.8/1.9 DeprecationWarning fixes. `#396 <https://github.com/darklow/django-suit/issues/396>`_ `#365 <https://github.com/darklow/django-suit/pull/365>`_ `#383 <https://github.com/darklow/django-suit/issues/383>`_
* [Feature] Django 1.8 show_change_link inlines parameter implementation. `#366 <https://github.com/darklow/django-suit/issues/366>`_
* [Fix] Fix issues related to Django 1.7 opts.module_name to opts.model_name transition. `#405 <https://github.com/darklow/django-suit/issues/405>`_
* [Fix] Add {% welcome-msg %} block around welcome message text. `#379 <https://github.com/darklow/django-suit/issues/379>`_


v0.2.13 (2015-04-22)
--------------------

* [Fix] Django 1.8 related fixes
* [Fix] Minor fixes and accepted PR: `#358 <https://github.com/darklow/django-suit/pull/358>`_ `#322 <https://github.com/darklow/django-suit/pull/322>`_ `#356 <https://github.com/darklow/django-suit/pull/356>`_ `#357 <https://github.com/darklow/django-suit/pull/357>`_ `#325 <https://github.com/darklow/django-suit/pull/325>`_ `#333 <https://github.com/darklow/django-suit/pull/333>`_ `#351 <https://github.com/darklow/django-suit/pull/351>`_ `#354 <https://github.com/darklow/django-suit/issues/354>`_ `#360 <https://github.com/darklow/django-suit/issues/360>`_
* [Tests] Fixed Django 1.8 and deprecated travis tests


v0.2.12 (2014-11-19)
--------------------

* [Fix] Raise an exception if inline fields are defined as tuple. `#302 <https://github.com/darklow/django-suit/pull/302>`_ [Thanks to @peterfarrell]
* [Fix] Fixes change_form first field focus. `#290 <https://github.com/darklow/django-suit/pull/290>`_ `#295 <https://github.com/darklow/django-suit/issues/295>`_ [Thanks to @cybersimon]
* [Fix] Django 1.7: Fix LinkedSelect if extra url args. `#310 <https://github.com/darklow/django-suit/issues/310>`_ `#294 <https://github.com/darklow/django-suit/issues/294>`_


v0.2.11 (2014-09-11)
--------------------

* [Fix] Django 1.7 compatibility fixes


v0.2.10 (2014-09-04)
--------------------

* [Fix] Fixes `#257 <https://github.com/darklow/django-suit/pull/257>`_ Fix for django-filer>=0.9.6 [Thanks to @mkutgt72]
* [Fix] Fixes `#266 <https://github.com/darklow/django-suit/pull/266>`_ Fixed issue where SortableGenericInlines would break whenever adding blank inline forms [Thanks to @sixthgear]
* [Fix] Fixes `#270 <https://github.com/darklow/django-suit/pull/270>`_ Use get_absolute_url() if the object has_absolute_url is True [Thanks to @stvbdn]


v0.2.9 (2014-06-20)
-------------------

* [Fix] Fixes `#240 <https://github.com/darklow/django-suit/issues/240>`_ Unable to save inline models with a FileField and "sortable" enabled.
* [Fix] Fixes `#232 <https://github.com/darklow/django-suit/issues/232>`_ Inline sortable error when model features only FKs


v0.2.8 (2014-04-22)
-------------------

* [Feature] Mark active search filters. Show label prefix in selected filter option. Closes `#207 <https://github.com/darklow/django-suit/issues/207>`_


v0.2.7 (2014-03-21)
-------------------


suit/widgets.py
from django.contrib.admin.widgets import AdminTimeWidget, AdminDateWidget
from django.forms import TextInput, Select, Textarea
from django.utils.safestring import mark_safe
from django import forms
from django.utils.translation import gettext as _
from django.templatetags.static import static

from suit import utils

django_version = utils.django_major_version()


class NumberInput(TextInput):
    """
    HTML5 Number input
    Left for backwards compatibility
    """
    input_type = 'number'


class HTML5Input(TextInput):
    """
    Supports any HTML5 input
    http://www.w3schools.com/html/html5_form_input_types.asp
    """

    def __init__(self, attrs=None, input_type=None):
        self.input_type = input_type
        super(HTML5Input, self).__init__(attrs)


#
class LinkedSelect(Select):
    """
    Linked select - Adds link to foreign item, when used with foreign key field
    """

    def __init__(self, attrs=None, choices=()):
        attrs = _make_attrs(attrs, classes="linked-select")
        super(LinkedSelect, self).__init__(attrs, choices)


class EnclosedInput(TextInput):
    """
    Widget for bootstrap appended/prepended inputs
    """

    def __init__(self, attrs=None, prepend=None, append=None):
        """
        For prepend, append parameters use string like %, $ or html
        """
        self.prepend = prepend
        self.append = append
        super(EnclosedInput, self).__init__(attrs=attrs)

    def enclose_value(self, value):
        """
        If value doesn't starts with html open sign "<", enclose in add-on tag
        """
        if value.startswith("<"):
            return value
        if value.startswith("icon-"):
            value = '<i class="%s"></i>' % value
        return '<span class="add-on">%s</span>' % value

    def render(self, name, value, attrs=None, renderer=None):
        if django_version < (2, 0):
            output = super(EnclosedInput, self).render(name, value, attrs)
        else:
            output = super(EnclosedInput, self).render(name, value, attrs, renderer)

        div_classes = []
        if self.prepend:
            div_classes.append('input-prepend')
            self.prepend = self.enclose_value(self.prepend)
            output = ''.join((self.prepend, output))
        if self.append:
            div_classes.append('input-append')
            self.append = self.enclose_value(self.append)
            output = ''.join((output, self.append))

        return mark_safe(
            '<div class="%s">%s</div>' % (' '.join(div_classes), output))


class AutosizedTextarea(Textarea):
    """
    Autosized Textarea - textarea height dynamically grows based on user input
    """

    def __init__(self, attrs=None):
        new_attrs = _make_attrs(attrs, {"rows": 2}, "autosize")
        super(AutosizedTextarea, self).__init__(new_attrs)

    @property
    def media(self):
        return forms.Media(js=[static("suit/js/jquery.autosize-min.js")])

    def render(self, name, value, attrs=None, renderer=None):
        if django_version < (2, 0):
            output = super(AutosizedTextarea, self).render(name, value, attrs)
        else:
            output = super(AutosizedTextarea, self).render(name, value, attrs, renderer)

        output += mark_safe(
            "<script type=\"text/javascript\">Suit.$('#id_%s').autosize();</script>"
            % name)
        return output


#
# Original date widgets with addition html
#
class SuitDateWidget(AdminDateWidget):
    def __init__(self, attrs=None, format=None):
        defaults = {'placeholder': _('Date:')[:-1]}
        new_attrs = _make_attrs(attrs, defaults, "vDateField input-small")
        super(SuitDateWidget, self).__init__(attrs=new_attrs, format=format)

    def render(self, name, value, attrs=None, renderer=None):
        if django_version < (1, 11):
            output = super(SuitDateWidget, self).render(name, value, attrs)
        else:
            output = super(SuitDateWidget, self).render(name, value, attrs, renderer)
        return mark_safe(
            '<div class="input-append suit-date">%s<span '
            'class="add-on"><i class="icon-calendar"></i></span></div>' %
            output)


class SuitTimeWidget(AdminTimeWidget):
    def __init__(self, attrs=None, format=None):
        defaults = {'placeholder': _('Time:')[:-1]}
        new_attrs = _make_attrs(attrs, defaults, "vTimeField input-small")
        super(SuitTimeWidget, self).__init__(attrs=new_attrs, format=format)

    def render(self, name, value, attrs=None, renderer=None):
        if django_version < (2, 0):
            output = super(SuitTimeWidget, self).render(name, value, attrs)
        else:
            output = super(SuitTimeWidget, self).render(name, value, attrs, renderer)
        return mark_safe(
            '<div class="input-append suit-date suit-time">%s<span '
            'class="add-on"><i class="icon-time"></i></span></div>' %
            output)


class SuitSplitDateTimeWidget(forms.SplitDateTimeWidget):
    """
    A SplitDateTime Widget that has some admin-specific styling.
    """

    def __init__(self, attrs=None):
        widgets = [SuitDateWidget, SuitTimeWidget]
        forms.MultiWidget.__init__(self, widgets, attrs)

    if django_version < (1, 11):
        def format_output(self, rendered_widgets):
            out_tpl = '<div class="datetime">%s %s</div>'
            return mark_safe(out_tpl % (rendered_widgets[0], rendered_widgets[1]))
    else:
        def render(self, name, value, attrs=None, renderer=None):
            output = super(SuitSplitDateTimeWidget, self).render(name, value, attrs, renderer)
            return mark_safe('<div class="datetime">%s</div>' % output)


def _make_attrs(attrs, defaults=None, classes=None):
    result = defaults.copy() if defaults else {}
    if attrs:
        result.update(attrs)
    if classes:
        result["class"] = " ".join((classes, result.get("class", "")))
    return result


suit/templatetags/suit_menu.py
import re
import warnings

import django
from django import template
from django.contrib import admin
from django.contrib.admin import AdminSite
from django.http import HttpRequest
from django.urls import resolve, reverse

from suit import utils
from suit.config import get_config

register = template.Library()

django_version = utils.django_major_version()

simple_tag = register.simple_tag


@simple_tag(takes_context=True)
def get_menu(context, request):
    """
    :type request: HttpRequest
    """
    if not isinstance(request, HttpRequest):
        return None

    # Try to get app list
    if hasattr(request, 'current_app'):
        # Django 1.8 uses request.current_app instead of context.current_app
        template_response = get_admin_site(request.current_app).index(request)
    else:
        try:
            template_response = get_admin_site(context.current_app).index(request)
        # Django 1.10 removed the current_app parameter for some classes and functions. 
        # Check the release notes.
        except AttributeError:
            template_response = get_admin_site(context.request.resolver_match.namespace).index(request)

    try:
        app_list = template_response.context_data['app_list']
    except Exception:
        return

    return Menu(context, request, app_list).get_app_list()


def get_admin_site(current_app):
    """
    Method tries to get actual admin.site class, if any custom admin sites
    were used. Couldn't find any other references to actual class other than
    in func_closer dict in index() func returned by resolver.
    """
    try:
        resolver_match = resolve(reverse('%s:index' % current_app))
        # Django 1.9 exposes AdminSite instance directly on view function
        if hasattr(resolver_match.func, 'admin_site'):
            return resolver_match.func.admin_site

        for func_closure in resolver_match.func.__closure__:
            if isinstance(func_closure.cell_contents, AdminSite):
                return func_closure.cell_contents
    except:
        pass
    return admin.site


class Menu(object):
    app_activated = False
    MULTIPLE_MODELS_RE = re.compile(r'([^*]*)[*]')

    def __init__(self, context, request, app_list):
        self.request = request
        self.app_list = app_list

        # Detect current app, if any
        try:
            self.ctx_app = context['app_label'].lower()
        except Exception:
            self.ctx_app = None

        # Get current model plural name, if any
        try:
            self.ctx_model_plural = context['opts'].verbose_name_plural.lower()
        except Exception:
            self.ctx_model_plural = None

        # Flatten all models from native apps
        self.all_models = [model for app in app_list for model in app['models']]

        # Init config variables
        self.init_config()

        super(Menu, self).__init__()

    def init_config(self):
        self.conf_exclude = get_config('MENU_EXCLUDE')
        self.conf_open_first_child = get_config('MENU_OPEN_FIRST_CHILD')
        self.conf_icons = get_config('MENU_ICONS')
        self.conf_menu_order = get_config('MENU_ORDER')
        self.conf_menu = get_config('MENU')

    def get_app_list(self):
        menu = None
        if self.conf_menu:
            menu = self.make_menu(self.conf_menu)
        elif self.conf_menu_order:
            menu = self.make_menu_from_old_format(self.conf_menu_order)
        else:
            menu = self.make_menu_from_native_only()

        # Add icons and match active
        if menu:
            self.activate_menu(menu)

        return menu

    def make_menu(self, config):
        menu = []
        if not isinstance(config, (tuple, list)):
            raise TypeError('Django Suit MENU config parameter must be '
                            'tuple or list. Got %s' % repr(config))
        for app in config:
            app = self.make_app(app)
            if app:
                menu.append(app)

        return menu

    def make_app(self, app_def):
        if isinstance(app_def, dict):
            app = app_def.copy()
        elif isinstance(app_def, str):
            if app_def == '-':
                app = self.make_separator()
            else:
                app = self.make_app_from_native(app_def)
        else:
            raise TypeError('MENU list item must be string or dict. Got %s'
                            % repr(app_def))
        if app:
            return self.process_app(app)

    def process_app(self, app):

        if 'app' in app:
            app = self.process_semi_native_app(app)

        if not app:
            return

        # Process icons
        self.process_icons(app)

        # Ensure required keys for template are set
        self.ensure_app_keys(app)

        # Exclude apps
        if self.app_is_excluded(app):
            return

        # Handle app permissions
        if self.app_is_forbidden(app):
            return

        # Process app models
        self.process_models(app)

        # Set link from child
        models = app.get('models', [])
        if self.conf_open_first_child and models:
            app['orig_url'] = app['url']
            app['url'] = models[0]['url']

        # Process absolute/named/model type urls
        app['url'] = self.process_url(app['url'], app)

        return app


    def app_is_forbidden(self, app):
        return app['permissions'] and \
               not self.user_has_permission(app['permissions'])

    def app_is_excluded(self, app):
        return self.conf_exclude and app['name'] in self.conf_exclude

    def process_icons(self, app):
        """
        If icon key is present but value is '' or None, set empty 'icon-'
        If key not found, try to set icon from SUIT_ICONS
        """
        if 'icon' in app:
            app['icon'] = app['icon'] or 'icon-'
        elif self.conf_icons and 'name' in app and \
                        app['name'] in self.conf_icons:
            app['icon'] = self.conf_icons[app['name']]

    def process_semi_native_app(self, app):
        """
        Process app defined as { app: 'app' }
        """
        app_from_native = self.make_app_from_native(app['app'])
        if app_from_native:
            del app['app']
            app_from_native.update(app)
            return app_from_native

    def make_menu_from_native_only(self):
        menu = []
        for app in self.app_list:
            app_name = ''
            app_url = app.get('app_url')
            if app_url:
                app_url_parts = app['app_url'].split('/')
                if len(app_url_parts) > 1:
                    app_name = app_url_parts[-2]
            app = self.convert_native_app(app, app_name)
            if app:
                app = self.process_app(app)
            if app:
                menu.append(app)

        return menu

    def make_app_from_native(self, app_name):
        app = self.find_native_app(app_name)
        if app:
            return self.convert_native_app(app, app_name)

    def find_native_app(self, app_name):
        for app in self.app_list:
            _app_name = app.get('app_label', app.get('name'))
            if _app_name.lower() == app_name:
                return app

    def convert_native_app(self, native_app, app_name):
        models = []
        native_models = native_app.get('models', {})
        if native_models:
            for model in native_models:
                models.append(self.convert_native_model(model, app_name))

        # Skip native apps with no models
        if not models:
            return

        return {
            'label': native_app['name'],
            'url': native_app['app_url'],
            'models': models,
            'name': app_name
        }

    def make_separator(self):
        return {
            'separator': True
        }

    def process_models(self, app):
        models = []
        models_def = app.get('models', [])
        for model_def in models_def:
            # multiple models may be returned
            models += self.make_models(model_def, app['name'])

        app['models'] = models

    def make_models(self, model_def, app_name):
        if not isinstance(model_def, str):
            model = self.make_model(model_def, app_name)
            return [model] if model else []
        match = self.MULTIPLE_MODELS_RE.match(model_def)
        if not match:
            model = self.make_model(model_def, app_name)
            return [model] if model else []
        prefix = match.group(1)
        prefix = self.get_model_name(app_name,prefix)
        return [
            m
            for m in [
                self.convert_native_model(native_model,app_name)
                for native_model in self.all_models
                if self.get_native_model_name(native_model).startswith(prefix)
            ]
            if m
        ]

    def make_model(self, model_def, app_name):
        if isinstance(model_def, dict):
            model = model_def.copy()
        elif isinstance(model_def, str):
            model = self.make_model_from_native(model_def, app_name)
        else:
            raise TypeError('MENU list item must be string or dict. Got %s'
                            % repr(model_def))
        if model:
            return self.process_model(model, app_name)

    def make_model_from_native(self, model_name, app_name):
        model = self.find_native_model(model_name, app_name)
        if model:
            return self.convert_native_model(model, app_name)

    def find_native_model(self, model_name, app_name):
        model_name = self.get_model_name(app_name, model_name)
        for native_model in self.all_models:
            if model_name == self.get_native_model_name(native_model):
                return native_model

    def model_is_excluded(self, model_name):
        return self.conf_exclude and model_name in self.conf_exclude

    def get_model_name(self, app_name, model_name):
        if app_name:
            app_name = app_name.lower()
        if model_name:
            model_name = model_name.lower()
        if '.' not in model_name:
            model_name = '%s.%s' % (app_name, model_name)
        return model_name

    def get_native_model_name(self, model):
        """
        Get model name by its last part of url
        """
        url_parts = self.get_native_model_url(model).rstrip('/').split('/')
        root_url_parts = reverse('admin:index').rstrip('/').split('/')
        return '.'.join(url_parts[len(root_url_parts):][:2])

    def convert_native_model(self, model, app_name):
        return {
            'label': model['name'],
            'url': self.get_native_model_url(model),
            'name': self.get_native_model_name(model),
            'app': app_name,
            'perms': model.get('perms',None),
            'add_url': model.get('add_url',None),
        }

    def get_native_model_url(self, model):
        return model.get('admin_url') or model.get('add_url', '')

    def process_model(self, model, app_name):
        if 'model' in model:
            model = self.process_semi_native_model(model, app_name)

        if model:
            self.ensure_model_keys(model)

            if 'app' in model and 'name' in model:
                model_name = self.get_model_name(model['app'], model['name'])

                if self.model_is_excluded(model_name):
                    return

            # Handle model permissions
            if self.model_is_forbidden(model):
                return

            # Detect if named url and convert it to absolute
            model['url'] = self.process_url(model['url'])

            return model

    def model_is_forbidden(self, model):
        return model['permissions'] and \
               not self.user_has_permission(model['permissions'])

    def process_semi_native_model(self, model, app_name):
        """
        Process app defined as { model: 'model' }
        """
        model_from_native = self.make_model_from_native(model['model'],
                                                        app_name)
        if model_from_native:
            del model['model']
            model_from_native.update(model)
            return model_from_native

    def ensure_app_keys(self, app):
        keys = ['label', 'url', 'icon', 'permissions', 'name', 'is_active',
                'blank']
        self.fill_keys(app, keys)

    def ensure_model_keys(self, model):
        keys = ['label', 'url', 'permissions', 'is_active', 'blank']
        self.fill_keys(model, keys)

    def fill_keys(self, dict, keys):
        for key in keys:
            if key not in dict:
                dict[key] = None

    def user_has_permission(self, perms):
        perms = perms if isinstance(perms, (list, tuple)) else (perms,)
        return self.request.user.has_perms(perms)

    def activate_menu(self, menu):
        for app in menu:

            # Make 'model' key as 'models' to unite activation logic
            if 'model' in app and not app['models']:
                app['models'] = [app['model']]

            # Activate models
            if app['models']:
                self.activate_models(app)

            # Mark as active by url match
            if not self.app_activated \
                and (self.request.path == app['url']
                     or self.request.path == app.get('orig_url')):
                app['is_active'] = self.app_activated = True

        if not self.app_activated:
            self.activate_menu_by_url(menu)

        # Last chance, try to activate by name
        if not self.app_activated:
            for app in menu:
                self.activate_models(app, match_by_name=True)

    def activate_models(self, app, match_by_name=False):
        for model in app['models']:
            if not match_by_name:
                # Mark as active by url or model plural name match
                model['is_active'] = self.request.path == model['url']
            else:
                model['is_active'] = self.ctx_model_plural == model[
                    'label'].lower()

            # Mark parent as active too
            if model['is_active'] and not self.app_activated:
                app['is_active'] = self.app_activated = True

    def activate_menu_by_url(self, menu):
        """
        If no active app/model is found in good/correct way, try to match
        by simple "startswith" in request path. Some apps doesn't provide
        nice app_label (django-filer for ex.) therefore this is the only way
        """
        for app in menu:
            for model in app['models']:
                if model['url'] and self.request.path.startswith(model['url']):
                    model['is_active'] = True
                    app['is_active'] = self.app_activated = True
                    break
            if self.app_activated:
                break

        # If still no active app found, match by app original url if any
        if not self.app_activated:
            for app in menu:
                orig_url = app.get('orig_url')
                if orig_url and self.request.path.startswith(orig_url):
                    app['is_active'] = self.app_activated = True

    def process_url(self, url, app=None):
        """
        Try to guess if it is absolute url or named
        """
        if url is None:
            return ''

        if not url or '/' in url:
            return url

        # Model link, ex: 'auth.user'
        if '.' in url:
            url_parts = url.split('.')
            model = self.make_model_from_native(url_parts[1], url_parts[0])
            if model:
                if app:
                    app['model'] = model
                return model['url']

        # Try to resolve as named url, ex: 'admin:index'
        try:
            return reverse(url)
        except:
            return url

    def make_menu_from_old_format(self, conf_order):
        import sys

        if 'test' not in sys.argv:
            warnings.warn(
                'Django Suit "MENU_ORDER" setting is deprecated. Use new "MENU"'
                ' key instead, see Documentation for new syntax.',
                DeprecationWarning)

        new_conf = []
        for order in conf_order:
            new_app = {}
            if isinstance(order, (tuple, list)):
                app_name = order[0]
                models_order = order[1] if len(order) > 1 else None
                if isinstance(app_name, str):
                    new_app['app'] = app_name
                elif isinstance(app_name, (tuple, list)):
                    mapping = ('label', 'url', 'icon', 'permissions')
                    for i, val in enumerate(app_name):
                        new_app[mapping[i]] = val
                if models_order and isinstance(models_order, (tuple, list)):
                    models = []
                    for model in models_order:
                        if isinstance(model, str):
                            models.append({'model': model})
                        elif isinstance(model, (list, tuple)):
                            mapping = ('label', 'url', 'permissions')
                            new_model = {}
                            for i, val in enumerate(model):
                                new_model[mapping[i]] = val
                            models.append(new_model)

                    new_app['models'] = models
            if new_app:
                new_conf.append(new_app)

        return self.make_menu(new_conf)
